include/pybind11/detail/internals.h

```diff
@@ -168,15 +173,31 @@ struct override_hash {
     }
 };
 
+using instance_map = std::unordered_multimap<const void *, instance *>;
+
+struct instance_map_shard {
+    std::mutex mutex;
+    instance_map registered_instances;
+    char padding[64 - (sizeof(std::mutex) + sizeof(instance_map)) % 64];
+};
+
 /// Internal data structure used to track registered instances and types.
 /// Whenever binary incompatible changes are made to this structure,
 /// `PYBIND11_INTERNALS_VERSION` must be incremented.
 struct internals {
+#if PYBIND11_INTERNALS_VERSION >= 6
+    std::mutex mutex;
+#endif
     // std::type_index -> pybind11's type information
     type_map<type_info *> registered_types_cpp;
     // PyTypeObject* -> base type_info(s)
     std::unordered_map<PyTypeObject *, std::vector<type_info *>> registered_types_py;
-    std::unordered_multimap<const void *, instance *> registered_instances; // void * -> instance*
+#if PYBIND11_INTERNALS_VERSION >= 6
+    std::unique_ptr<instance_map_shard[]> instance_shards; // void * -> instance*
+    size_t instance_shards_mask;
+#else
+    instance_map registered_instances; // void * -> instance*
```

Comment:
I assume the ABI would remain unchanged in a non-free-threaded build? It's not 100% clear just from looking at this change.